apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc/states/encrypted_txs.rs

```diff
@@ -0,0 +1,96 @@
+use std::marker::PhantomData;
+
+use super::super::{AllocStatus, BlockSpaceAllocator};
+use super::{
+    BuildingEncryptedTxBatch, FillingRemainingSpace, NextStateImpl, State,
+    WithEncryptedTxs, WithoutEncryptedTxs,
+};
+
+impl State for BlockSpaceAllocator<BuildingEncryptedTxBatch<WithEncryptedTxs>> {
+    #[inline]
+    fn try_alloc(&mut self, tx: &[u8]) -> AllocStatus {
+        self.encrypted_txs.try_dump(tx)
+    }
+
+    #[inline]
+    fn try_alloc_batch<'tx, T>(&mut self, txs: T) -> AllocStatus
+    where
+        T: IntoIterator<Item = &'tx [u8]> + 'tx,
+    {
+        self.encrypted_txs.try_dump_all(txs)
+    }
+}
+
+impl NextStateImpl
+    for BlockSpaceAllocator<BuildingEncryptedTxBatch<WithEncryptedTxs>>
+{
+    type Next = BlockSpaceAllocator<FillingRemainingSpace<WithEncryptedTxs>>;
+
+    #[inline]
+    fn next_state_impl(self) -> Self::Next {
+        next_state(self)
+    }
+}
+
+impl State
+    for BlockSpaceAllocator<BuildingEncryptedTxBatch<WithoutEncryptedTxs>>
+{
+    #[inline]
+    fn try_alloc(&mut self, tx: &[u8]) -> AllocStatus {
+        AllocStatus::Rejected {
+            tx_len: tx.len() as u64,
+            space_left: 0,
+        }
+    }
+
+    #[inline]
+    fn try_alloc_batch<'tx, T>(&mut self, _txs: T) -> AllocStatus
+    where
+        T: IntoIterator<Item = &'tx [u8]> + 'tx,
+    {
+        AllocStatus::Rejected {
+            // arbitrary `tx_len` value; doesn't really matter what we
+            // choose here, as long as it's greater than zero
+            tx_len: u64::MAX,
+            space_left: 0,
+        }
```

Comment:
This is kind of a code smell though, relating to the comment at https://github.com/anoma/namada/pull/791/files#diff-91ec5a00b45ee4a93117e944d2cdd7c12774c3e0ae852489ea296a6055a77836R231 i.e. `AllocStatus` isn't an appropriate return type for this `try_alloc_batch` fn (maybe it would be cleaner to just let callers do iteration if they need to)